In [14]:
import pandas as pd
import numpy as np

In [2]:
data = pd.read_csv('wine_equality.csv')
data.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            8.3              0.90         0.02             2.4      0.083   
1            7.4              0.70         0.00             1.9      0.076   
2            7.8              0.88         0.00             2.6      0.098   
3            7.8              0.76         0.04             2.3      0.092   
4           11.2              0.28         0.56             1.9      0.075   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 15.0                  45.0   0.9932  4.02       0.63   
1                 11.0                  34.0   0.9978  3.51       0.56   
2                 25.0                  67.0   0.9968  3.20       0.68   
3                 15.0                  54.0   0.9970  3.26       0.65   
4                 17.0                  60.0   0.9980  3.16       0.58   

   alcohol  quality  
0      9.6        6  
1      9.4        5  
2      9.8        5  
3      9.8        5  
4      9.8        6

In [3]:
data.isnull().sum()

fixed acidity           0
volatile acidity        0
citric acid             0
residual sugar          0
chlorides               0
free sulfur dioxide     0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

In [4]:
#sklearn mannual implementation of the code
#first we peform collinearity for all the columns and remove one of the columns if the collineraity is greater than 0.5
# Calculate correlation matrix
correlation_matrix = data.corr()
correlation_matrix.to_csv('correlation.csv')

In [5]:
correlation_matrix

fixed acidity  volatile acidity  citric acid  \
fixed acidity              1.000000         -0.255800     0.671364   
volatile acidity          -0.255800          1.000000    -0.553136   
citric acid                0.671364         -0.553136     1.000000   
residual sugar             0.114777          0.001787     0.143582   
chlorides                  0.093706          0.061092     0.203793   
free sulfur dioxide       -0.153793         -0.010598    -0.060879   
total sulfur dioxide      -0.113181          0.076309     0.035551   
density                    0.667324          0.019536     0.365867   
pH                        -0.678552          0.239011    -0.541761   
sulphates                  0.183005         -0.260848     0.312739   
alcohol                   -0.061651         -0.202980     0.110448   
quality                    0.124041         -0.389421     0.225878   

                      residual sugar  chlorides  free sulfur dioxide  \
fixed acidity               0.114777   0.093706            -0.153793   
volatile acidity            0.001787   0.061092            -0.010598   
citric acid                 0.143582   0.203793            -0.060879   
residual sugar              1.000000   0.055615             0.187053   
chlorides                   0.055615   1.000000             0.005567   
free sulfur dioxide         0.187053   0.005567             1.000000   
total sulfur dioxide        0.203030   0.047403             0.667667   
density                     0.355006   0.200522            -0.021823   
pH                         -0.085374  -0.263566             0.069679   
sulphates                   0.005537   0.371266             0.051666   
alcohol                     0.042115  -0.221053            -0.069355   
quality                     0.013703  -0.128925            -0.050676   

                      total sulfur dioxide   density        pH  sulphates  \
fixed acidity                    -0.113181  0.667324 -0.678552   0.183005   
volatile acidity                  0.076309  0.019536  0.239011  -0.260848   
citric acid                       0.035551  0.365867 -0.541761   0.312739   
residual sugar                    0.203030  0.355006 -0.085374   0.005537   
chlorides                         0.047403  0.200522 -0.263566   0.371266   
free sulfur dioxide               0.667667 -0.021823  0.069679   0.051666   
total sulfur dioxide              1.000000  0.071243 -0.066188   0.042951   
density                           0.071243  1.000000 -0.344450   0.148536   
pH                               -0.066188 -0.344450  1.000000  -0.195836   
sulphates                         0.042951  0.148536 -0.195836   1.000000   
alcohol                          -0.205594 -0.494634  0.202049   0.093657   
quality                          -0.185101 -0.175244 -0.056065   0.251332   

                       alcohol   quality  
fixed acidity        -0.061651  0.124041  
volatile acidity     -0.202980 -0.389421  
citric acid           0.110448  0.225878  
residual sugar        0.042115  0.013703  
chlorides            -0.221053 -0.128925  
free sulfur dioxide  -0.069355 -0.050676  
total sulfur dioxide -0.205594 -0.185101  
density              -0.494634 -0.175244  
pH                    0.202049 -0.056065  
sulphates             0.093657  0.251332  
alcohol               1.000000  0.475830  
quality               0.475830  1.000000

<h3><div>removing one of the columns having the correlation less than 0.5</div></h3>

In [6]:
highly_correlated = set()
for i in range(len(correlation_matrix.columns)):
    for j in range(i):
        if abs(correlation_matrix.iloc[i, j]) > 0.5:
            colname_i = correlation_matrix.columns[i]
            colname_j = correlation_matrix.columns[j]
            highly_correlated.add((colname_i, colname_j))
columns_to_drop = set()
for col1, col2 in highly_correlated:
    columns_to_drop.add(col2)

data_after_removal = data.drop(columns=columns_to_drop, axis=1)
print("the columns which we are dropping are :\n\t",columns_to_drop,"\n")
print("the remaining columns are :\n\t",data_after_removal.columns)

the columns which we are dropping are :
	 {'volatile acidity', 'free sulfur dioxide', 'citric acid', 'fixed acidity'} 

the remaining columns are :
	 Index(['residual sugar', 'chlorides', 'total sulfur dioxide', 'density', 'pH',
       'sulphates', 'alcohol', 'quality'],
      dtype='object')


<h3><div>the eliminated columns are Citric_acid , Fixed_acidity , Free_sulfur_dioxide , Volatile_acidity</div></h3>

In [7]:
dar = data_after_removal
dar.head()

residual sugar  chlorides  total sulfur dioxide  density    pH  sulphates  \
0             2.4      0.083                  45.0   0.9932  4.02       0.63   
1             1.9      0.076                  34.0   0.9978  3.51       0.56   
2             2.6      0.098                  67.0   0.9968  3.20       0.68   
3             2.3      0.092                  54.0   0.9970  3.26       0.65   
4             1.9      0.075                  60.0   0.9980  3.16       0.58   

   alcohol  quality  
0      9.6        6  
1      9.4        5  
2      9.8        5  
3      9.8        5  
4      9.8        6

In [8]:
dar.isnull().sum()

residual sugar          0
chlorides               0
total sulfur dioxide    0
density                 0
pH                      0
sulphates               0
alcohol                 0
quality                 0
dtype: int64

<h3><div>there is no preprocessing required as the data does not have any null values</div></h3>

In [9]:
dar.head()

residual sugar  chlorides  total sulfur dioxide  density    pH  sulphates  \
0             2.4      0.083                  45.0   0.9932  4.02       0.63   
1             1.9      0.076                  34.0   0.9978  3.51       0.56   
2             2.6      0.098                  67.0   0.9968  3.20       0.68   
3             2.3      0.092                  54.0   0.9970  3.26       0.65   
4             1.9      0.075                  60.0   0.9980  3.16       0.58   

   alcohol  quality  
0      9.6        6  
1      9.4        5  
2      9.8        5  
3      9.8        5  
4      9.8        6

<h3>first let me use logisitic regression with threshold of quality greater than 6 as good and less than 6 as bad<br></h3>

In [10]:
from sklearn.model_selection import train_test_split

In [11]:
X = dar.iloc[:,0:-1]
X

residual sugar  chlorides  total sulfur dioxide  density    pH  \
0                2.4      0.083                  45.0  0.99320  4.02   
1                1.9      0.076                  34.0  0.99780  3.51   
2                2.6      0.098                  67.0  0.99680  3.20   
3                2.3      0.092                  54.0  0.99700  3.26   
4                1.9      0.075                  60.0  0.99800  3.16   
...              ...        ...                   ...      ...   ...   
1595             2.0      0.090                  44.0  0.99490  3.45   
1596             2.2      0.062                  51.0  0.99512  3.52   
1597             2.3      0.076                  40.0  0.99574  3.42   
1598             2.0      0.075                  44.0  0.99547  3.57   
1599             3.6      0.067                  42.0  0.99549  3.39   

      sulphates  alcohol  
0          0.63      9.6  
1          0.56      9.4  
2          0.68      9.8  
3          0.65      9.8  
4          0.58      9.8  
...         ...      ...  
1595       0.58     10.5  
1596       0.76     11.2  
1597       0.75     11.0  
1598       0.71     10.2  
1599       0.66     11.0  

[1600 rows x 7 columns]

In [12]:
dar['quality'].value_counts()

quality
5    681
6    639
7    199
4     53
8     18
3     10
Name: count, dtype: int64

In [13]:
dar['quality'] = dar['quality'].map({3 : 'bad', 4 :'bad', 5: 'bad',
                                      6: 'good', 7: 'good', 8: 'good'}) 
dar['quality']

0       good
1        bad
2        bad
3        bad
4       good
        ... 
1595     bad
1596    good
1597    good
1598     bad
1599    good
Name: quality, Length: 1600, dtype: object

In [14]:
from sklearn.preprocessing import LabelEncoder

le = LabelEncoder()

dar['quality'] = le.fit_transform(dar['quality'])

dar['quality'].value_counts()

y = dar['quality']
y = np.array(y)
y

array([1, 0, 0, ..., 1, 0, 1])

In [15]:
x_train,x_test,y_train,y_test = train_test_split(X,y,test_size = 0.2,random_state = 0)

In [16]:
def sigmoid(z):
    return 1 / (1 + np.exp(-z))
def logistic_regression(X, y, lr=0.01, epochs=1000):
    m, n = X.shape
    theta = np.zeros(n) 

    for epoch in range(epochs):
        z = np.dot(X, theta)
        h = sigmoid(z)
        error = h - y

        gradient = np.dot(X.T, error) / m
        theta -= lr * gradient

    return theta

theta = logistic_regression(x_train, y_train)
print(theta)

[-0.03657666 -0.0231719  -0.03610832 -0.10959049 -0.36507474  0.1403035
  0.48903785]


In [17]:
count = 0
predicted = []
for i in range(x_test.shape[0]):
    predicted.append(np.dot(x_test.iloc[i],theta))
for i in range(x_test.shape[0]):
    if predicted[i] > 0.6:
        predicted[i] = 1
    else:
        predicted[i] = 0
    if y_test[i] - predicted[i] == 0 :
        count = count + 1
accuracy = (count/x_test.shape[0])*100
print("testing_accuracy : ", accuracy)

testing_accuracy :  58.75


In [18]:
from sklearn.metrics import confusion_matrix

In [19]:
print(confusion_matrix(y_test,predicted))

[[ 36 125]
 [  7 152]]


<h3>we got an accuracy of 58.75% in logistic regression<br></h3>

In [15]:
#removing correlated columns and setting it as a new file dar.csv
dar.to_csv('dar.csv', sep=',', encoding='utf-8', index=False)

In [21]:
count = 0
train_predicted = []
for i in range(x_train.shape[0]):
    train_predicted.append(np.dot(x_train.iloc[i],theta))
for i in range(x_train.shape[0]):
    if train_predicted[i] > 0.6:
        train_predicted[i] = 1
    else:
        train_predicted[i] = 0
    if y_train[i] - train_predicted[i] == 0 :
        count = count + 1
accuracy = (count/x_train.shape[0])*100
print("training_accuracy : ", accuracy)

training_accuracy :  63.046875
